### DISCUSSION TITLE
LCD Display: SysEx, string and ghost of a string

### DISCUSSION THREAD

**Discussion by eqcq** [FortySevenEffects/arduino_midi_library#254] [Q&A]
Hi and thank you so much for this library that make my life brighter everyday :)
I'm listening midi sysex message reception on my esp32 with that code :

```
void OnMidiSysEx(byte* data, unsigned length) {
 digitalWrite(ONBOARD_LED, HIGH);
 byte tempByte=0;
 char tempMessage=' ';
 String message="";
 for (int i = 2; i < data[length - 2]; i++) {
   tempByte = data[i];
   tempMessage = (char)tempByte;
   message += String(tempMessage);
 }
 Ecran.fillScreen(TFT_BLACK);
 Ecran.drawString(message, 0, 0, 2);
}

```
The message is the good one and I can store the string I need and display it. But it seems that my string doesn't initizialize correctly each time I send a new message :

for info I start the "for" loop at 2 because the first 2 bits are the header of the message so we don't read it, same for the length -2 because we do not read the last byte.

If I send the following sysex (where F0 11 and F7 are the system bytes): f0 11 65 6c 65 70 68 61 6e 74 f7 my screen displays elephant

But if after I send a shorter one : f0 11 72 61 74 f7 for "rat", my screen display « rathant » as if a chunk of my string was still stored somewhere

I've test the length of my byte array in order to controle the length was changing and it's ok.

Sorry, it's must be easy but I don't see my mistake.

Thanks for all!

**Comment by franky47** [comment]
This has little to do with SysEx, and more with how LCD screens store data in their memory.

To reduce the amount of cycles per character (which shortens its life), it keeps an internal memory of what was displayed before, and only overwrites with incoming data.

To have consistent writes, you have two options:
- Generate a string the length of your LCD display (eg: 16 characters wide), filled with space characters (0x20), then write your message onto that string and send it entirely to the LCD screen.
- Send a "clear" message to the LCD before writing your shorter message. This may cause some flicker.

**Reply by eqcq** [reply]
Ok I've found the mistake

change 
`message+=String(tempMessage)`
for
`message.concat(String(tempMessage));`

thanks for the support

**Reply by eqcq** [reply]
Sorry, finally there was still a mistake I've corrected by replacing : 

` byte tempByte=0;
 char tempMessage=' ';
 String message="";
 for (int i = 2; i < data[length - 2]; i++) {
   tempByte = data[i];
   tempMessage = (char)tempByte;
   message += String(tempMessage);
 }`

by 

`    std::string myString((char*)pData, length);
  String message = myString.c_str();`
